File: Java/src/Ton/Ton.java
/*
MIT License

Copyright (c) 2018 Váraljai Péter

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

package Ton;

import java.io.*;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Ton implements Serializable, Cloneable
{
    HashMap<String,Ton> keyvalues;

    //*/// Constructors

    public Ton()
    {
        keyvalues = new HashMap<>();
    }

    public Ton(String key)
    {
        keyvalues = new HashMap<>();
        this.put(key,new Ton());
    }

    public Ton(String key, String value)
    {
        this.put(key,value);
    }

    //*/// Normal HasMap manipulation

    public boolean containsKey(String key)
    {
        return keyvalues.containsKey(key);
    }

    public Ton get(String key)
    {
        return keyvalues.get(key);
    }

    //Private cos no need to create a self contained mess.
    private void put(String key, Ton ton)
    {
        keyvalues.put(key,ton);
    }

    public Set<String> keySet()
    {
        return keyvalues.keySet();
    }

    public Collection<Ton> values()
    {
        return keyvalues.values();
    }

    //*/// Extra HasMap manipulation

    // If the key already exist don't overwrite, also has no return value
    public void put(String key)
    {
        if (this.containsKey(key))
            return;
        else
            this.put(key,new Ton());
    }

    // If the key already exist put another value, also has no return value
    public void put(String key, String value)
    {
        if (this.containsKey(key))
            this.get(key).put(value);
        else
            this.put(key,new Ton(value));
    }

    public void putAll(String key, Collection<String> values)
    {
        for (String v : values)
        {
            this.put(key,v);
        }
    }

    //*/// Statics

    static public Ton parsefromFile(String fileName) throws IOException
    {
        FileReader fileReader = new FileReader(fileName);
        BufferedReader bufferedReader = new BufferedReader(fileReader);
        try
        {
            return parsefromString(bufferedReader);
        }
        finally
        {
            bufferedReader.close();
        }

    }

    static public Ton parsefromString(String raw) throws IOException
    {
        InputStream is = new ByteArrayInputStream(raw.getBytes());
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(is));
        return parsefromString(bufferedReader);
    }

    static public Ton parsefromString(BufferedReader bufferedReader) throws IOException
    {
        Ton ret = new Ton();

        Stack<Integer> stack = new Stack<>();
        HashMap<Integer,String> hash = new HashMap<>();
        Integer last = -1;

        String line = null;
        while((line = bufferedReader.readLine()) != null)
        {
            Pattern p = Pattern.compile("^([\t ]*)([^\t ]*.*)$"); // (tabs or spaces)(others)
            Matcher m = p.matcher(line);
            if (m.find())
            {
                Integer t = m.group(1).length(); //Number of indentation chars
                String value = m.group(2);
                //System.out.println(t+" "+value);

                hash.put(t,value);

                if (t>last)
                {
                    stack.push(t);
                }
                else if (t<last)
                {
                    while (stack.size()>0 && stack.lastElement()>t)
                    {
                        stack.pop();
                    }
                }
                last = t;

                Ton a = ret;
                ArrayList<Integer> list = new ArrayList(stack);
                for (Integer tt : list)
                {
                    String key = hash.get(tt);
                    if (!a.containsKey(key))
                    {
                        a.put(key);
                    }
                    a = a.get(key);
                    //System.out.println(hash.get(tt));
                }
                //System.out.println();

            }
        }
        return ret;
    }

    static private String ton2String(Ton ton,int lvl)
    {
        StringBuilder ret = new StringBuilder();
        String t = new String(new char[lvl]).replace("\0", "\t"); // Until String::repeate

        for(String key : ton.keySet() )
        {
            ret.append(t+key+"\n"+ton2String(ton.get(key),lvl+1)); //RECURSION!!!
        }

        return ret.toString();
    }

    @Override
    public String toString()
    {
        return ton2String(this,0);
    }
}

File: CPP/Ton/Ton.cpp
/*
MIT License

Copyright (c) 2018 Váraljai Péter

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#include "Ton/Ton.h"

//*/// Member

Ton::Ton()
{}

Ton::Ton(std::string key)
{
    this->insert(key);
}

Ton::~Ton() 
{
    //std::cout << this << "\n";
    for(auto it = keyvalues.begin(); it != keyvalues.end(); ++it)
    {
        delete it->second;
    }
    keyvalues.clear();
}

//*/// Iterators
void Ton::getKeys(std::vector<std::string> &keys)
{
    for(auto it = keyvalues.begin(); it != keyvalues.end(); ++it)
    {
        keys.push_back(it->first);
    }
}

//*/// Element access, Modifiers

Ton* Ton::at(std::string key)
{
    std::unordered_map<std::string,Ton*>::const_iterator got = keyvalues.find(key);
    if (got!=keyvalues.end()) return got->second;

    return nullptr;
}

void Ton::insert(std::string key)
{
    // If the key already exist don't overwrite
    if (!this->containsKey(key)) keyvalues[key]=new Ton();  
}

void Ton::insert(std::string key, std::string value)
{
    // If the key already exist insert another value
    if (!this->containsKey(key)) 
        keyvalues[key]=new Ton(value);
    else
        this->at(key)->insert(value);
}

//*/// Element lookup

bool Ton::containsKey(std::string key)
{
    return ( keyvalues.find(key)!=keyvalues.end() );
}

//*/// Static in

Ton* Ton::parseFromFile(std::string fileName)
{
    std::istream* s = new std::ifstream(fileName);
    Ton* ret = Ton::parseFromString(*s);
    delete s;
    return ret;
}

Ton* Ton::parseFromString(std::string raw)
{
    std::istream* s = new std::stringstream(raw);
    Ton* ret = Ton::parseFromString(*s);
    delete s;
    return ret;
}

Ton* Ton::parseFromString(std::istream &raw)
{
    Ton* ret = new Ton();

    std::vector<int> stack;
    std::unordered_map<int,std::string> hash;
    int last = -1;

    std::string line;
    while (std::getline(raw, line))
    {
        std::regex re("^([\t ]*)([^\t ].*)$");
        std::smatch match;
        std::regex_search(line, match, re);
        if (match.size() > 1) 
        {
            int t = match.str(1).length();
            std::string value = match.str(2);
            //std::cout << t << " " << value << "\n";

            hash[t]=value;

            if (t>last)
            {
                stack.push_back(t);
            }
            else if (t<last)
            {
                while (stack.size()>0 && stack.back()>t)
                {
                    stack.pop_back();
                }
            }
            last = t;

            Ton* a = ret;
            for(int tt : stack)
            {
                std::string key = hash[tt];
                a->insert(key);
                a = a->at(key);
                //std::cout << key << "\n";
            }
            //std::cout << "\n";
        }
    }

    return ret;
}

//*/// Static out

std::string Ton::ton2string(Ton* ton,int lvl)
{
    std::string ret;
    std::string t = std::string(lvl, '\t');

    for(auto it = ton->keyvalues.begin(); it != ton->keyvalues.end(); ++it)
    {
        ret+=(t+it->first+"\n"+ton2string(it->second,lvl+1)); //RECURSION!!!
    }

    return ret;
}

std::string Ton::toString()
{
    return Ton::ton2string(this,0);
}

